<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Fei Qin" />


<title>SCRIP simulation for scRNA-seq data</title>


<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">SCRIP simulation for scRNA-seq data</h1>
<h4 class="author">Fei Qin</h4>
<h4 class="date">Last updated: 11/16/2021</h4>



<div id="introduction-to-scrip-method" class="section level1">
<h1>1. Introduction to SCRIP method</h1>
<p>SCRIP proposed two frameworks based on Gamma-Poisson and Beta-Gamma-Poisson distribution for simulating scRNA-seq data. Both Gamma-Poisson and Beta-Gamma-Poisson distribution model the over dispersion of scRNA-seq data. Specifically, Beta-Gamma-Poisson model was used to model bursting effect. The dispersion was accurately simulated by fitting the mean-BCV dependency using generalized additive model (GAM). Other key characteristics of scRNA-seq data including library size, zero inflation and outliers were also modeled by SCRIP. With its flexible modeling, SCRIP enables various application for different experimental designs and goals including DE analysis, clustering analysis, trajectory-based analysis and bursting analysis</p>
</div>
<div id="installation" class="section level1">
<h1>2. Installation</h1>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;splatter&quot;</span>)</a>
<a class="sourceLine" id="cb1-2" title="2"></a>
<a class="sourceLine" id="cb1-3" title="3"><span class="kw">library</span>(devtools)</a>
<a class="sourceLine" id="cb1-4" title="4"><span class="kw">install_github</span>(<span class="st">&quot;thecailab/SCRIP&quot;</span>)</a></code></pre></div>
</div>
<div id="quick-start" class="section level1">
<h1>3. Quick start</h1>
<p>Assuming you already have a count matrix for scRNA-seq data, and you want to simulation data based on it. Only a few steps are needed to creat a simulation data using SCRIP.</p>
<p>A dataset from Xin data is used for example.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="kw">library</span>(splatter)</a>
<a class="sourceLine" id="cb2-2" title="2"><span class="kw">library</span>(SCRIP)</a>
<a class="sourceLine" id="cb2-3" title="3"> </a>
<a class="sourceLine" id="cb2-4" title="4"><span class="kw">data</span>(acinar.data)</a>
<a class="sourceLine" id="cb2-5" title="5">params &lt;-<span class="st"> </span><span class="kw">splatEstimate</span>(acinar.data)</a></code></pre></div>
<pre><code>## $start.arg
## $start.arg$shape
## [1] 0.833088
## 
## $start.arg$rate
## [1] 0.09357466
## 
## 
## $fix.arg
## NULL
## 
## $start.arg
## $start.arg$meanlog
## [1] 9.415808
## 
## $start.arg$sdlog
## [1] 1.034692
## 
## 
## $fix.arg
## NULL
## 
## $start.arg
## $start.arg$meanlog
## [1] 4.719586
## 
## $start.arg$sdlog
## [1] 0.7954047
## 
## 
## $fix.arg
## NULL</code></pre>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1">sim_trend &lt;-<span class="st">  </span><span class="kw">SCRIPsimu</span>(<span class="dt">data=</span>acinar.data, <span class="dt">params=</span>params, <span class="dt">mode=</span><span class="st">&quot;GP-trendedBCV&quot;</span>)</a>
<a class="sourceLine" id="cb4-2" title="2">sim_trend</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 1000 80 
## metadata(13): Params method ... batch.facScale bcv.shrink
## assays(5): BatchCellMeans BaseCellMeans CellMeans TrueCounts counts
## rownames(1000): Gene1 Gene2 ... Gene999 Gene1000
## rowData names(4): Gene BaseGeneMean OutlierFactor GeneMean
## colnames(80): Cell1 Cell2 ... Cell79 Cell80
## colData names(3): Cell Batch ExpLibSize
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):</code></pre>
<div style="page-break-after: always;"></div>
</div>
<div id="single-cell-type-simulation" class="section level1">
<h1>4 Single cell type simulation</h1>
<div id="parameter-estimation" class="section level2">
<h2>4.1 Parameter estimation</h2>
<p>SCRIP utlized the estimation strategy from splatter, but also provided more parameters (Fold change, dropout rates, library size, BCV degree of freedom) to serve different experimental designs (i.e. Simulation for differential expression analysis, clustering analysis and trajectory analysis). Detailed description about other parameters will be shown in other sections of this document.</p>
</div>
<div id="simulation" class="section level2">
<h2>4.2 Simulation</h2>
<p>The default mode in SCRIP for simulation is “GP-trendedBCV”. You can also choose other modes (“GP-commonBCV”, “BGP-commonBCV”,“BP”, “BGP-trendedBCV”) in the SCRIPsimu() function. For single cell type simulation, you have to set the “method” as “single”, which was default in SCRIPsimu() function.</p>
<div id="gp-commonbcv" class="section level3">
<h3>4.2.1 GP-commonBCV</h3>
<p>GP-commonBCV is the model used by splatter. GP-commonBCV applied the Gamma-Poisson mixture model with mean-BCV dependency fitted by a common BCV across genes.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1"><span class="co">########################### GP-commonBCV model/Splatter ##########################</span></a>
<a class="sourceLine" id="cb6-2" title="2"><span class="co">##################################################################################</span></a>
<a class="sourceLine" id="cb6-3" title="3">sim_GPcommon &lt;-<span class="st">  </span><span class="kw">SCRIPsimu</span>(<span class="dt">data=</span>acinar.data, <span class="dt">params=</span>params, <span class="dt">mode=</span><span class="st">&quot;GP-commonBCV&quot;</span>)</a>
<a class="sourceLine" id="cb6-4" title="4">sim_GPcommon</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 1000 80 
## metadata(13): Params method ... batch.facScale bcv.shrink
## assays(5): BatchCellMeans BaseCellMeans CellMeans TrueCounts counts
## rownames(1000): Gene1 Gene2 ... Gene999 Gene1000
## rowData names(4): Gene BaseGeneMean OutlierFactor GeneMean
## colnames(80): Cell1 Cell2 ... Cell79 Cell80
## colData names(3): Cell Batch ExpLibSize
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):</code></pre>
</div>
<div id="gp-trendedbcv" class="section level3">
<h3>4.2.2 GP-trendedBCV</h3>
<p>GP-trendedBCV is the major model of SCIRP. which used the Gamma-Poisson mixture model with mean-BCV dependency fitted by GAM.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1"><span class="co">############################### GP-trendedBCV model ##############################</span></a>
<a class="sourceLine" id="cb8-2" title="2"><span class="co">##################################################################################</span></a>
<a class="sourceLine" id="cb8-3" title="3">sim_GPtrend &lt;-<span class="st">  </span><span class="kw">SCRIPsimu</span>(<span class="dt">data=</span>acinar.data, <span class="dt">params=</span>params, <span class="dt">mode=</span><span class="st">&quot;GP-trendedBCV&quot;</span>)</a></code></pre></div>
</div>
<div id="bp" class="section level3">
<h3>4.2.3 BP</h3>
<p>BP is the model used for simulating bursting effect using Beta-Poisson mixture distributionwithout considering BCV effect.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1"><span class="co">############################### BP-commonBCV model ##############################</span></a>
<a class="sourceLine" id="cb9-2" title="2"><span class="co">##################################################################################</span></a>
<a class="sourceLine" id="cb9-3" title="3">sim_BP &lt;-<span class="st">  </span><span class="kw">SCRIPsimu</span>(<span class="dt">data=</span>acinar.data, <span class="dt">params=</span>params, <span class="dt">mode=</span><span class="st">&quot;BP&quot;</span>)</a></code></pre></div>
<pre><code>## $start.arg
## $start.arg$shape1
## [1] 0.1792874
## 
## $start.arg$shape2
## [1] 1.938204
## 
## 
## $fix.arg
## NULL</code></pre>
</div>
<div id="bgp-commonbcv" class="section level3">
<h3>4.2.4 BGP-commonBCV</h3>
<p>BP-commonBCV is the model used for simulating bursting effect with Beta-Gamma-Poisson mixture distribution. The mean-BCV dependency was fitted by a common BCV across genes.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1"><span class="co">############################### BP-commonBCV model ##############################</span></a>
<a class="sourceLine" id="cb11-2" title="2"><span class="co">##################################################################################</span></a>
<a class="sourceLine" id="cb11-3" title="3">sim_BGPcommon &lt;-<span class="st">  </span><span class="kw">SCRIPsimu</span>(<span class="dt">data=</span>acinar.data, <span class="dt">params=</span>params, <span class="dt">mode=</span><span class="st">&quot;BGP-commonBCV&quot;</span>)</a></code></pre></div>
<pre><code>## $start.arg
## $start.arg$shape1
## [1] 0.1792874
## 
## $start.arg$shape2
## [1] 1.938204
## 
## 
## $fix.arg
## NULL</code></pre>
</div>
<div id="bgp-trendedbcv" class="section level3">
<h3>4.2.5 BGP-trendedBCV</h3>
<p>BP-trendedBCV is the model used for simulating bursting effect with Beta-Gamma-Poisson mixture distribution. The mean-BCV dependency was fitted by a GAM.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1"><span class="co">############################### BP-trendedBCV model ##############################</span></a>
<a class="sourceLine" id="cb13-2" title="2"><span class="co">##################################################################################</span></a>
<a class="sourceLine" id="cb13-3" title="3">sim_BGPtrend &lt;-<span class="st">  </span><span class="kw">SCRIPsimu</span>(<span class="dt">data=</span>acinar.data, <span class="dt">params=</span>params, <span class="dt">mode=</span><span class="st">&quot;BGP-trendedBCV&quot;</span>)</a></code></pre></div>
<pre><code>## $start.arg
## $start.arg$shape1
## [1] 0.1792874
## 
## $start.arg$shape2
## [1] 1.938204
## 
## 
## $fix.arg
## NULL</code></pre>
<div style="page-break-after: always;"></div>
</div>
</div>
</div>
<div id="group-simulation" class="section level1">
<h1>5 Group simulation</h1>
<p>Group simulation is useful for studying different experimental conditions, especially for differential expression (DE) analysis. To serve different applications in scRNA-seq analysis, SCRIP provides flexible simulation. It can simulate scRNA-seq data with different parameters from multiple cell groups (i.e. cell types), which is useful for evaluating the detection of global characteristics such as clustering. It also allows simulation of group difference in a single cell group, which is useful for evaluating typical DE analysis methods.</p>
<div id="basic-group-simulation" class="section level2">
<h2>5.1 Basic group simulation</h2>
<p>DEGs were simulated using multiplicative differential expression factors from a log-normal distribution with parameters including number of genes (nGenes), the path-specific proportion of DE genes (de.prob), the proportion of down-regulated DE genes (de.downProb), DE location factor (de.facLoc) and DE scale factor (de.facScale).</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1">sim.SCRIP2 &lt;-<span class="st"> </span><span class="kw">SCRIPsimu</span>(<span class="dt">data=</span>acinar.data, <span class="dt">params=</span>params, <span class="dt">method=</span><span class="st">&quot;groups&quot;</span>, </a>
<a class="sourceLine" id="cb15-2" title="2">                        <span class="dt">batchCells=</span><span class="dv">300</span>, <span class="dt">group.prob =</span> <span class="kw">c</span>(<span class="fl">0.25</span>, <span class="fl">0.25</span>, <span class="fl">0.25</span>, <span class="fl">0.25</span>), </a>
<a class="sourceLine" id="cb15-3" title="3">                        <span class="dt">de.prob =</span> <span class="kw">c</span>(<span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>), </a>
<a class="sourceLine" id="cb15-4" title="4">                        <span class="dt">de.downProb =</span> <span class="kw">c</span>(<span class="fl">0.5</span>, <span class="fl">0.5</span>, <span class="fl">0.5</span>, <span class="fl">0.5</span>),</a>
<a class="sourceLine" id="cb15-5" title="5">                        <span class="dt">de.facLoc =</span> <span class="kw">c</span>(<span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">0.4</span>, <span class="fl">0.5</span>), </a>
<a class="sourceLine" id="cb15-6" title="6">                        <span class="dt">de.facScale=</span><span class="kw">c</span>(<span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>)) </a></code></pre></div>
</div>
<div id="group-simulation-with-batch-effect" class="section level2">
<h2>5.2 Group simulation with batch effect</h2>
<p>Batch effect factors are also generated from a log-normal distribution with parameters including batchCells, batch.facLoc and batch.facScale.</p>
<p>batchCells: number of cells for each batch  <br />
batch.facLoc: Batch location factor in log-normal distribution for batch factor  <br />
batch.facScale: Batch scale factor in log-normal distribution for batch factor  </p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" title="1">sim.SCRIP3 &lt;-<span class="st"> </span><span class="kw">SCRIPsimu</span>(<span class="dt">data=</span>acinar.data, <span class="dt">params=</span>params, <span class="dt">method=</span><span class="st">&quot;groups&quot;</span>, </a>
<a class="sourceLine" id="cb16-2" title="2">                        <span class="dt">batchCells=</span><span class="kw">c</span>(<span class="dv">150</span>, <span class="dv">150</span>), </a>
<a class="sourceLine" id="cb16-3" title="3">                        <span class="dt">batch.facLoc =</span> <span class="kw">c</span>(<span class="fl">0.1</span>, <span class="fl">0.1</span>), </a>
<a class="sourceLine" id="cb16-4" title="4">                        <span class="dt">batch.facScale =</span> <span class="kw">c</span>(<span class="fl">0.1</span>, <span class="fl">0.1</span>), </a>
<a class="sourceLine" id="cb16-5" title="5">                        <span class="dt">group.prob =</span> <span class="kw">c</span>(<span class="fl">0.25</span>, <span class="fl">0.25</span>, <span class="fl">0.25</span>, <span class="fl">0.25</span>), </a>
<a class="sourceLine" id="cb16-6" title="6">                        <span class="dt">de.prob =</span> <span class="kw">c</span>(<span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>), </a>
<a class="sourceLine" id="cb16-7" title="7">                        <span class="dt">de.downProb =</span> <span class="kw">c</span>(<span class="fl">0.5</span>, <span class="fl">0.5</span>, <span class="fl">0.5</span>, <span class="fl">0.5</span>),</a>
<a class="sourceLine" id="cb16-8" title="8">                        <span class="dt">de.facLoc =</span> <span class="kw">c</span>(<span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">0.4</span>, <span class="fl">0.5</span>), </a>
<a class="sourceLine" id="cb16-9" title="9">                        <span class="dt">de.facScale=</span><span class="kw">c</span>(<span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>)) </a></code></pre></div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
